rust/template/d3log/src/error.rs

```diff
@@ -0,0 +1,144 @@
+// this file follows the apparently established pattern of defining a local error type and defining From
+// so as to implicity coerce all the errors thrown by '?'
+//TODOs:
+//  write a macro for the common to_string case below
+//  refactor(?) async_error so it doesn't require its users pull in a bunch of dependencies
+//  consider adding support for nesting
+//  macro for record field extraction case
+//  consider adding support for source location if thats possible
+
+use core::convert::Infallible;
+use core::fmt;
+use core::fmt::Display;
+use std::ffi::NulError;
+use std::str::Utf8Error;
+use tokio::task::JoinError;
+
+#[derive(Debug)]
+pub struct Error {
+    contents: String,
+    line: Option<u64>,
+    functionname: Option<String>,
+    filename: Option<String>,
+    parent: Box<Option<Error>>,
+}
+
+#[macro_export]
+macro_rules! function {
+    () => {{
+        fn f() {}
+        fn type_name_of<T>(_: T) -> &'static str {
+            std::any::type_name::<T>()
+        }
+        let name = type_name_of(f);
+        &name[..name.len() - 3]
+    }};
+}
+
+impl Error {
+    pub fn new(s: String) -> Error {
+        Error {
+            contents: s,
+            line: None,
+            functionname: None,
+            filename: None,
+            parent: Box::new(None),
+        }
+    }
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(&self.contents)
+    }
+}
+
+// can we really just alpha this ? string is different(?).
+impl From<Utf8Error> for Error {
+    fn from(err: Utf8Error) -> Self {
+        Error::new(err.to_string())
+    }
+}
+
+impl From<std::io::Error> for Error {
+    fn from(err: std::io::Error) -> Self {
+        Error::new(err.to_string())
+    }
+}
+
+impl From<NulError> for Error {
+    fn from(err: NulError) -> Self {
+        Error::new(err.to_string())
+    }
+}
+
+impl From<Infallible> for Error {
+    fn from(_err: Infallible) -> Self {
+        panic!("can't happen");
```

Comment:
```suggestion
    fn from(err: Infallible) -> Self {
        match err {}
```
Neat little trick with uninhabited types